nodes/hoppers.lua

```diff
@@ -18,33 +18,49 @@ local function get_hopper_formspec(pos)
 	return formspec
 end
 
+-- unfortunately param2 overrides are needed for side hoppers even in the non-single-craftable-item case
+-- because they are literally *side* hoppers - their spouts point to the side rather than to the front, so
+-- the default item_place_node orientation code will not orient them pointing toward the selected surface.
 local hopper_on_place = function(itemstack, placer, pointed_thing, node_name)
-	local pos  = pointed_thing.under
-	local pos2 = pointed_thing.above
-	local x = pos.x - pos2.x
-	local z = pos.z - pos2.z
-
 	local returned_stack, success
-	-- unfortunately param2 overrides are needed for side hoppers even in the non-single-craftable-item case
-	-- because they are literally *side* hoppers - their spouts point to the side rather than to the front, so
-	-- the default item_place_node orientation code will not orient them pointing toward the selected surface.
-	if x == -1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 0)
-	elseif x == 1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 2)
-	elseif z == -1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side")  then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 3)
-	elseif z == 1 and (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
-		returned_stack, success = minetest.item_place_node(ItemStack("hopper:hopper_side"), placer, pointed_thing, 1)
+
+	if not (hopper.config.single_craftable_item or node_name == "hopper:hopper_side") then
+		returned_stack, success = minetest.item_place_node(itemstack, placer, pointed_thing)
+		return returned_stack
+	end
+
+	local pointed_pos  = pointed_thing.under
+	local hopper_pos   = pointed_thing.above
+
+	local param2_by_offset = {
+		[vector.new(-1, 0, 0):to_string()] = 0,
```

Comment:
FYI: `minetest.hash_node_position` might be faster. There are no hash collisions within the current world limits (-2¹⁵ to 2¹⁵-1).

Reply:
> FYI: minetest.hash_node_position might be faster. There are no hash collisions within the current world limits (-2¹⁵ to 2¹⁵-1).

Write the offset as the node position? It sounds very unreliable. 

> EDIT: Also this table could be moved outside this function to calculate it exactly once (optimization).

I think LuaJIT is very good at optimization and will handle it without my help.

Reply:
> Write the offset as the node position? It sounds very unreliable.

Not really. Lua uses `double` types internally, which has a mantissa (fraction) part of 52 bits. Hence the 16 * 3 = 48 is guaranteed to be represented without precision loss.

> I think LuaJIT is very good at optimization and will handle it without my help.

It's so time-critical code, so either way is fine I think.